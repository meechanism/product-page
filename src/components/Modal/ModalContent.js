import React from 'react';
import ReactDOM from 'react-dom';
import FocusTrap from 'focus-trap-react';
import styled from 'styled-components';

import Colors from '../../styles/colors';
import MediaQuery from '../../styles/media-queries';

const ModalWrapper = styled.aside.attrs({
  tag: 'aside',
  role: 'dialog',
  tabIndex: -1,
  'aria-modal': true
})`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 10;
  transform: translateZ(0);
  background-color: rgba(0, 0, 0, 0.5);
`;

const CloseButton = styled.button.attrs((props) => ({
  'aria-label': 'Close Modal',
  'aria-labelledby': 'close-modal'
}))`
  position: absolute;
  top: 0;
  right: 0;
  padding: 0.5em;
  line-height: 1;
  background: ${Colors.trueWhite};
  border: 0;
  box-shadow: 0;
  cursor: pointer;

  &:hover {
    svg {
      stroke: ${Colors.primary};
    }
  }
`;

const ModalArea = styled.div`
  background-color: ${Colors.trueWhite};
  box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.1);
  border-radius: 2px;
  position: fixed;
  width: 100%;
  overflow-y: auto;
  overflow-x: hidden;
  -webkit-overflow-scrolling: touch;

  left: 50%;
  top: 50%;
  height: auto;
  transform: translate(-50%, -50%);
  max-width: 30em;
  max-height: calc(100% - 1em);

  ${MediaQuery.small`
    width: 100%;
  `};
`;

const CloseIcon = styled.svg`
  width: 25px;
  height: 25px;
  fill: transparent;
  stroke: black;
  stroke-linecap: round;
  stroke-width: 2;
`;

const CloseVisual = styled.span`
  border: 0 !important;
  clip: rect(0 0 0 0) !important;
  height: 1px !important;
  margin: -1px !important;
  overflow: hidden !important;
  padding: 0 !important;
  position: absolute !important;
  width: 1px !important;
  white-space: nowrap !important;
`;

const ModalContent = ({
  buttonRef,
  onClickOutside,
  onKeyDown,
  modalRef,
  closeModal,
  content
}) => {
  return ReactDOM.createPortal(
    <FocusTrap>
      <ModalWrapper onClick={onClickOutside} onKeyDown={onKeyDown}>
        <ModalArea ref={modalRef}>
          <CloseButton ref={buttonRef} onClick={closeModal}>
            <CloseVisual id="close-modal">Close</CloseVisual>
            <CloseIcon viewBox="0 0 40 40">
              <path d="M 10,10 L 30,30 M 30,10 L 10,30" />
            </CloseIcon>
          </CloseButton>
          <div>{content}</div>
        </ModalArea>
      </ModalWrapper>
    </FocusTrap>,
    document.body
  );
};

export default ModalContent;
