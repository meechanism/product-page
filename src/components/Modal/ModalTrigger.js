import React from 'react';
import Button from '../Button';

export const onClick = (showModal) => (e) => {
  e.preventDefault();
  showModal();
};

const ModalTrigger = ({ buttonRef, showModal, triggerText }) => (
  <Button secondary ref={buttonRef} onClick={onClick(showModal)}>
    {triggerText}
  </Button>
);

export default ModalTrigger;
