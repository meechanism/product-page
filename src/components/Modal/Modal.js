import React, { useState } from 'react';
import ModalContent from './ModalContent';
import ModalTrigger from './ModalTrigger';

export const onKeyDown = (closeFn) => (event) => {
  if (event.keyCode === 27) {
    closeFn();
  }
};

export const toggleScrollLock = () => {
  document.querySelector('html').classList.toggle('scroll-lock');
};

export const showModal = (showFn, scrollLock, closeButton) => () => {
  showFn(true);
  closeButton && closeButton.focus();
  scrollLock();
};

export const closeModal = (showFn, scrollLock, triggerButton) => () => {
  showFn(false);
  triggerButton && triggerButton.focus();
  scrollLock();
};

export const onClickOutside = (closeFn, modal) => (event) => {
  if (modal && modal.contains(event.target)) return;
  closeFn();
};

export const Modal = ({ modalProps, modalContent }) => {
  const [isShown, setIsShown] = useState(false);
  const [modal, setModal] = useState(null);
  const [triggerButton, setTriggerButton] = useState(null);
  const [closeButton, setCloseButton] = useState(null);

  const closeFn = closeModal(setIsShown, toggleScrollLock, triggerButton);
  return (
    <>
      <ModalTrigger
        showModal={showModal(setIsShown, toggleScrollLock, closeButton)}
        buttonRef={(n) => setTriggerButton(n)}
        triggerText={modalProps.triggerText}
      />
      {isShown ? (
        <ModalContent
          modalRef={(n) => setModal(n)}
          buttonRef={(n) => setCloseButton(n)}
          closeModal={closeFn}
          content={modalContent}
          onKeyDown={onKeyDown(closeFn)}
          onClickOutside={onClickOutside(closeFn, modal)}
        />
      ) : null}
    </>
  );
};

export default Modal;
