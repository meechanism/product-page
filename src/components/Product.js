import React, { useState } from 'react';
import styled from 'styled-components';
import he from 'he';

import Modal from './Modal/Modal';
import ProductAssets from './ProductAssets';
import Colors from '../styles/colors';

const ProductWrapper = styled.a`
  display: block;
  color: ${Colors.black};
  line-height: 1.2rem;
  margin: 0;
  border-radius: 2px;
  position: relative;
  padding: 2rem;
  border: 1px solid transparent;
  &:hover {
    border: 1px solid ${Colors.gray};
  }
`;

const Img = styled.img`
  border-radius: 2px;
  width: 100%;
`;

const Name = styled.p`
  margin: 1rem 0;
  font-size: 0.9rem;
`;

const Price = styled.span`
  font-size: 0.9rem;
`;

const ImgWrapper = styled.div`
  position: relative;
`;

const HoverOverlay = styled.div`
  display: ${(props) => (props.show ? 'block' : 'none')};
  position: absolute;
  left: 0;
  bottom: 0rem;
  width: 100%;
`;

const saveQuicklook = (save, val) => () => save(val);

const Product = ({ product }) => {
  const [quickLook, setQuickLook] = useState(false);
  const { id, name, thumbnail, priceRange, hero, images } = product;
  const { selling } = priceRange;

  const decodedName = he.decode(name);

  return (
    <ProductWrapper href={`/details/${id}`}>
      <ImgWrapper
        onMouseOver={saveQuicklook(setQuickLook, true)}
        onMouseLeave={saveQuicklook(setQuickLook, false)}>
        <Img src={thumbnail.href} alt={name} />
        <HoverOverlay show={quickLook}>
          <Modal
            modalProps={{
              triggerText: 'Quicklook'
            }}
            modalContent={
              <ProductAssets
                compact
                name={decodedName}
                hero={hero}
                images={images}
              />
            }
          />
        </HoverOverlay>
      </ImgWrapper>
      <Name>{decodedName}</Name>
      <Price>{`$${selling.low} - $${selling.high}`}</Price>
    </ProductWrapper>
  );
};

export default Product;
