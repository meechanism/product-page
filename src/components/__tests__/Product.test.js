import React from 'react';
import he from 'he';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import Product, {
  saveQuicklook,
  Name,
  Price,
  Img,
  HoverOverlay
} from '../Product';

import Modal from '../Modal/Modal';

import productData from '../../data/products';

describe('Product component', () => {
  const testProduct = productData.groups[0];
  it('should render', () => {
    const output = shallow(<Product product={testProduct} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should trigger callback when invoked', () => {
    const mockFn = jest.fn();
    const testVal = 'linen';
    saveQuicklook(mockFn, testVal)();
    expect(mockFn).toHaveBeenCalledWith(testVal);
  });

  it('should render decoded name', () => {
    const output = shallow(<Product product={testProduct} />);
    const name = output.find(Name);
    expect(name.exists()).toBeTruthy();
    expect(name.text()).toBe(he.decode(testProduct.name));
  });

  it('should render price range', () => {
    const output = shallow(<Product product={testProduct} />);
    const price = output.find(Price);
    expect(price.exists()).toBeTruthy();
    const selling = testProduct.priceRange.selling;
    expect(price.text()).toBe(`$${selling.low} - $${selling.high}`);
  });

  it('should render thumbnail', () => {
    const output = shallow(<Product product={testProduct} />);
    const img = output.find(Img);
    expect(img.exists()).toBeTruthy();
    expect(img.props().src).toBe(testProduct.thumbnail.href);
  });

  describe('hover overlay', () => {
    it('should render hover overlay', () => {
      const output = shallow(<Product product={testProduct} />);
      const overlay = output.find(HoverOverlay);
      expect(overlay.exists()).toBeTruthy();
    });

    it('should render quicklook button', () => {
      const output = shallow(<Product product={testProduct} />);
      const modal = output.find(Modal);
      expect(modal.exists()).toBeTruthy();
    });

    it('should render quicklook button style with hidden state', () => {
      const tree = renderer.create(<HoverOverlay show={false} />).toJSON();
      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('display', 'none');
    });

    it('should render quicklook button style with shown state', () => {
      const tree = renderer.create(<HoverOverlay show={true} />).toJSON();
      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('display', 'block');
    });
  });
});
