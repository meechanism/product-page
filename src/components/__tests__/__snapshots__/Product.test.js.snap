// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component hover overlay should render quicklook button style with hidden state 1`] = `
.c0 {
  display: none;
  position: absolute;
  left: 0;
  bottom: 0rem;
  width: 100%;
}

<div
  className="c0"
/>
`;

exports[`Product component hover overlay should render quicklook button style with shown state 1`] = `
.c0 {
  display: block;
  position: absolute;
  left: 0;
  bottom: 0rem;
  width: 100%;
}

<div
  className="c0"
/>
`;

exports[`Product component should render 1`] = `
<styled.a
  href="/details/belgian-linen-embroidery-duvet-cover-shams-slate-white-b3024"
>
  <styled.div
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <styled.img
      alt="Belgian Linen Ladder Stripe Embroidery Duvet Cover &amp; Shams - Slate/White"
      src="https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-s-m.jpg"
    />
    <styled.div
      show={false}
    >
      <Modal
        modalContent={
          <ProductAssets
            compact={true}
            hero={
              Object {
                "alt": "",
                "height": 363,
                "href": "https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-s-m.jpg",
                "meta": "",
                "rel": "hero",
                "size": "m",
                "width": 363,
              }
            }
            images={
              Array [
                Object {
                  "alt": "",
                  "height": 363,
                  "href": "https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-6-m.jpg",
                  "meta": "",
                  "rel": "althero",
                  "size": "m",
                  "width": 363,
                },
                Object {
                  "alt": "",
                  "height": 363,
                  "href": "https://www.westelm.com/weimgs/ab/images/wcm/products/202003/0012/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-m.jpg",
                  "meta": "",
                  "rel": "althero",
                  "size": "m",
                  "width": 363,
                },
              ]
            }
            name="Belgian Linen Ladder Stripe Embroidery Duvet Cover & Shams - Slate/White"
          />
        }
        modalProps={
          Object {
            "triggerText": "Quicklook",
          }
        }
      />
    </styled.div>
  </styled.div>
  <styled.p>
    Belgian Linen Ladder Stripe Embroidery Duvet Cover & Shams - Slate/White
  </styled.p>
  <styled.span>
    $44 - $329
  </styled.span>
</styled.a>
`;
